internal/archive/tar.go

```diff
@@ -0,0 +1,185 @@
+// Package archive contains functionality to create tar archives
+package archive
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	ignore "github.com/sabhiram/go-gitignore"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/shipa-corp/ketch/internal/errors"
+)
+
+const (
+	shipaIgnoreFile = ".shipaignore"
+)
+
+type fileIgnoreList struct {
+	ign *ignore.GitIgnore
+}
+
+func createFileIgnoreList(ignoreFile string) (*fileIgnoreList, error) {
+	var il fileIgnoreList
+	ign, err := ignore.CompileIgnoreFile(ignoreFile)
+	if err != nil {
+		if !os.IsExist(err) {
+			return &il, nil
+		}
+		return nil, err
+	}
+	il.ign = ign
+	return &il, nil
+}
+
+func (fil fileIgnoreList) ignore(testFile string) bool {
+	if fil.ign == nil {
+		return false
+	}
+	return fil.ign.MatchesPath(testFile)
+}
+
+type tarOptions struct {
+	dirs       []string
+	files      []string
+	workingDir string
+}
+
+type Option func(options *tarOptions)
+
+// IncludeDirs defines directories to be included in tarball. Directories are relative to working directory.
+func IncludeDirs(dir ...string) Option {
+	return func(options *tarOptions) {
+		options.dirs = append(options.dirs, dir...)
+	}
+}
+
+// Defines files to be included in tarball. File paths are relative to the working directory.
+// Files included here override .shipaignore
+func IncludeFiles(file ...string) Option {
+	return func(options *tarOptions) {
+		options.files = append(options.files, file...)
+	}
+}
+
+// WithWorkingDirectory set working or root directory, of this is not set it defaults to os.Getwd
+func WithWorkingDirectory(path string) Option {
+	return func(options *tarOptions) {
+		options.workingDir = path
+	}
+}
+
+// Create a tar file named archiveFile. Use includeFiles and includeDirs to define the files to
+// be incorporated into the archive file. If no files or directories are specified the archive will be built
+// rooted at the current directory. withWorkingDirectory is used to set working directory and defaults
+// to the current working directory. The archival process can be instructed to exclude certain files by placing a
+// special file .shipaignore in the working directory.  It uses pattern matches like .gitignore such the files
+// matching the patterns are excluded from the archive. Files specified with includeFiles override .shipaignore and
+// are added to the archive.
+func Create(archiveFile string, inputs ...Option) error {
+	var options tarOptions
+	var err error
+	if options.workingDir, err = os.Getwd(); err != nil {
+		return errors.Wrap(err, "get working dir failed")
+	}
+	for _, input := range inputs {
+		input(&options)
+	}
+	// if no directories or files are specified, default to current directory
+	if len(options.dirs) == 0 && len(options.files) == 0 {
+		options.dirs = []string{"."}
+	}
+
+	// change to working directory
+	currentWd, err := os.Getwd()
```

Comment:
not sure if you need to call `os.Getwd()` multiple times here if [options.workingDir](https://github.com/shipa-corp/ketch/pull/50/files#diff-f0d23faf1e41e68a9840cedd68384283a0fd8beb383eb40299fa5b21e7be604eR85) gets the same value.